src/vos/vos_pool.c

```diff
@@ -885,15 +885,45 @@ vos_pool_open(const char *path, uuid_t uuid, unsigned int flags, daos_handle_t *
 	return vos_pool_open_metrics(path, uuid, flags, NULL, poh);
 }
 
-void
-vos_pool_features_set(daos_handle_t poh, uint64_t feats)
+int
+vos_pool_upgrade(daos_handle_t poh, uint32_t version)
 {
-	struct vos_pool	*pool;
+	struct vos_pool    *pool;
+	struct vos_pool_df *pool_df;
+	int                 rc = 0;
 
 	pool = vos_hdl2pool(poh);
 	D_ASSERT(pool != NULL);
 
-	pool->vp_feats |= feats;
+	pool_df = pool->vp_pool_df;
+
+	if (version == pool_df->pd_version)
+		return 0;
+
+	D_ASSERTF(version > pool_df->pd_version && version <= POOL_DF_VERSION,
```

Comment:
Could this happen if downgrade happen?

Reply:
I asked Johann if we support downgrade and he said no

Reply:
That said, if you prefer we return an error, I can do that.

Reply:
I think we had some codes to return error when pool service up for downgrade, i am not sure if vos_pool_upgrade() will be called before that or after that, maybe before, so it might trigger assertion if people try to downgrade daos.

Reply:
Ultimately, there is no way to reverse enabling the aggregation feature

Reply:
I spoke offline with @wangshilong.  We determined that this assertion can never actually happen under normal circumstances because the pool_upgrade can only be called on a pool that is open and an upgraded pool can't be opened in an older daos system.